### DISCUSSION TITLE
Batching draw calls with DrawIndex

### DISCUSSION THREAD

**Discussion by Myrindd** [gpuweb/gpuweb#4630] [Q&A] [answered]
I am currently porting a webgl2 engine to webgpu, but the engine uses DrawIndex (gl_DrawID) extensively in order to optimize the rendering of scenes with a lot of unique meshes with unique materials. Neither multidraw nor DrawIndex seems to be supported right now; is DrawIndex something we can expect to come soon?
Otherwise, since push-constants are also absent, is there a way to render the meshes without switching bind groups between each draw call?
For me, the motivation for moving to webgpu was reducing CPU overhead, but that would seem to defeat the whole point...

**Comment by kainino0x** [comment]
> Neither multidraw nor DrawIndex seems to be supported right now; is DrawIndex something we can expect to come soon?

Possibly as part of multi-draw-indirect #1354/#2315 but it will be some time before that gets added, I think.

Can you use dynamic offsets to change the binding offset between draw calls? It still needs a setBindGroup call but you can do it without creating tons of GPUBindGroup objects. See [`hasDynamicOffset`](https://gpuweb.github.io/gpuweb/#dictdef-gpubufferbindinglayout) and the `dynamicOffsets` parameter of [`setBindGroup()`](https://gpuweb.github.io/gpuweb/#gpubindingcommandsmixin-setbindgroup)

Also consider using [render bundles](https://gpuweb.github.io/gpuweb/#gpurenderbundleencoder) to reduce the number of JS calls you need to make each frame - if you are rendering the same thing frame-over-frame then you can bundle together everything that you would have done in one multidraw call in WebGL, and "replay" that each frame.

**Reply by Myrindd** [reply, marked as answer]
Thank you for the advice; dynamic offsets sounds like a good solution to limit how many bind groups are created; but still require manually changing offset between draw calls, which I would like to avoid through instancing.
Instead, I am currently investigating the use of instance index offsets (firstInstance option in the draw command). It will still require multiple draw calls (one per mesh) - at least until multidraw is implemented.
